internal/controller/atlasthirdpartyintegrations/setup.go

```diff
@@ -0,0 +1,196 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/controller/atlas"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/controller/reconciler"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/indexer"
+	akov2next "github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/nextapi/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/translation/project"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/translation/thirdpartyintegration"
+	ctrlstate "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/state"
+	mckpredicate "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/predicate"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/ratelimit"
+)
+
+var integrationPredicates = predicate.ResourceVersionChangedPredicate{}
+
+type AtlasThirdPartyIntegrationHandler struct {
+	ctrlstate.StateHandler[akov2next.AtlasThirdPartyIntegration]
+	reconciler.AtlasReconciler
+	deletionProtection bool
+}
+
+func NewAtlasThirdPartyIntegrationsReconciler(
+	c cluster.Cluster,
+	atlasProvider atlas.Provider,
+	deletionProtection bool,
+	logger *zap.Logger,
+	globalSecretRef client.ObjectKey,
+	reapplySupport bool,
+) *ctrlstate.Reconciler[akov2next.AtlasThirdPartyIntegration] {
+	handler := &AtlasThirdPartyIntegrationHandler{
+		AtlasReconciler: reconciler.AtlasReconciler{
+			Client:          c.GetClient(),
+			AtlasProvider:   atlasProvider,
+			Log:             logger.Named("controllers").Named("AtlasThirdPartyIntegration").Sugar(),
+			GlobalSecretRef: globalSecretRef,
+		},
+		deletionProtection: deletionProtection,
+	}
+	return ctrlstate.NewStateReconciler(
+		handler,
+		ctrlstate.WithCluster[akov2next.AtlasThirdPartyIntegration](c),
+		ctrlstate.WithReapplySupport[akov2next.AtlasThirdPartyIntegration](reapplySupport),
+	)
+}
+
+// For prepares the controller for its target Custom Resource; AtlasThirdPartyIntegration
+func (r *AtlasThirdPartyIntegrationHandler) For() (client.Object, builder.Predicates) {
+	return &akov2next.AtlasThirdPartyIntegration{}, builder.WithPredicates(integrationPredicates)
+}
+
+func (h *AtlasThirdPartyIntegrationHandler) SetupWithManager(mgr ctrl.Manager, rec reconcile.Reconciler, opts ...ctrlstate.SetupManagerOption) error {
+	h.Client = mgr.GetClient()
+	obj := &akov2next.AtlasThirdPartyIntegration{}
+	builder := controllerruntime.NewControllerManagedBy(mgr).
+		For(
+			obj,
+			ctrlrtbuilder.WithPredicates(
+				predicate.Or(
+					mckpredicate.AnnotationChanged("mongodb.com/reapply-period"),
+					predicate.GenerationChangedPredicate{},
+				),
+				mckpredicate.IgnoreDeletedPredicate[client.Object](),
+			),
+		).
+		Watches(
+			&akov2.AtlasProject{},
+			handler.EnqueueRequestsFromMapFunc(h.integrationForProjectMapFunc()),
+			builder.WithPredicates(integrationPredicates),
```

Comment:
the global variable `integrationPredicates` obfuscates what predicates we're having so reading through it makes it hard to reason about what predicates are actually applied. I suggest using `predicate.GenerationChangedPredicate{}` and `predicate.ResourceVersionChangedPredicate{}` to make it explicit and readable to the user.

In addition you risk faulty behaviour in case controller-runtime decides to have internal state inside the predicate struct. I'd be careful reusing it.